README.md document improvement (#49)
poggit/libasynql: README.md

# libasynql <img src="https://poggit.pmmp.io/ci.badge/poggit/libasynql/libasynql" align="right"/>
Asynchronous SQL access library for PocketMine plugins.

> ## Why should I use this library and what asynchronous means?

Reviewer: ```suggestion
## Why should I use this library and what does asynchronous mean?
```

> When excuting a SQL query at the main thread, **there will be a delay made** by MySQL for communicating between the database server or Sqlite for interacting with the file system. And the delay will completely block the main thread and **causes lag to the server**.

Reviewer: ```suggestion
When executing a SQL query on the main thread, **there will be a delay** to wait for the MySQL server or SQLite for interacting with the file system. The delay will block the main thread and **cause lag to the server**.
```

Author: Alright understable, I will use "on threads" instead of "at threads"!

> Libasynql uses **different threads for excuting the queries** so the main thread will not be affect by the lag!

Reviewer: ```suggestion
Libasynql uses **different threads for executing the queries** so the main thread will not lag!
```

Author: Alright understandable! I will use "on" instead of "at".

> *Look in [here](https://github.com/SOF3/pmmp-wilderness/wiki/PocketMine-Plugin-Development-FAQ#what-is-threading-does-it-make-the-server-faster) as well if you want to learn more about threadding.*

Reviewer: ```suggestion
*Look in [here](https://github.com/SOF3/pmmp-wilderness/wiki/PocketMine-Plugin-Development-FAQ#what-is-threading-does-it-make-the-server-faster) as well if you want to learn more about threading.*
```

### Creating SQL files
In the resources file, create one file for each SQL dialect you are supporting, e.g. `resources/sqlite.sql` and `resources/mysql.sql`.

#### Do I save the SQL files to the plugin data folder?

> No, you don't have to save / copy the SQL file to the plugin data folder. The content in the files can already be read by libasynql when they are in the resources folder

Reviewer: ```suggestion
No, you don't have to copy the SQL files to the plugin data folder. The files are read by libasynql from the resources folder directly.
```

They have their respective methods in DataConnector: `executeGeneric`, `executeChange`, `executeInsert`, `executeSelect`. They require the same parameters:

- The name of the prepared statement

> - The variables for the query, in the form of an associative array "variable name (without colon in the start)" => value

Reviewer: ```suggestion
- The variables for the query, in the form of an associative array "variable name (without the leading colon)" => value
```

###### `bool` default
`true`, `on`, `yes` or `1` will result in true. Other values, as long as there is something, will result default false. (If there is nothing, the variable will not be optional)

#### Example of using variables
##### SQL file
```sql
-- #{ example
-- #    { insert
-- # 	  :foo string
-- # 	  :bar int
INSERT INTO example(
	foo_column
	bar_column
) VALUES (

> 	:foo

Reviewer: ```suggestion
	:foo,
```

	:bar
);
-- #    }
-- #    { select
-- # 	  :foo string
-- # 	  :bar int
SELECT * FROM example
WHERE foo_column = :foo

> LIMIT :bar;

Reviewer: Have you tested if this works? I'm not completely sure if variables work with LIMIT.

Author: # Yes

Author: # Worked

-- #    }
-- #}
```
##### Code
```php
// Example of using variable in insert statements

> $this->database->excuteInsert("example.insert", ["foo" => "sample text", "bar" => 123]);

Reviewer: ```suggestion
$this->database->executeInsert("example.insert", ["foo" => "sample text", "bar" => 123]);
```

// Example of using variable in select statements

> $this->database->executeSelect("example.select", ["foo" => "sample text", "bar" => 1]);

Reviewer: This does nothing if you don't handle its result.

Author: Imma make it echo the bar column

Author: # Added

```

### Query text
Query text is not a command, but the non-commented part between the start and end commands of a query declaration.

Variables are interpolated in query text using the `:var` format. Note that libasynql uses a homebrew algorithm for identifying the variable positions, so they might be inaccurate.
```sql
-- #{ query.declarartion
SELECT * FROM example;
-- The line above is a query text
-- #}
```

## Things to beware

> ### Common mistakes

Reviewer: ```suggestion
### Race condition
```

```php
public $foo = 'bar';

public function setFoo() : void {
	$this->foo = 'foo';
}

public function getFoo() : string {
	return $this->foo;
}
```

```php
$this->database->executeGeneric("common.mistake.asynchronous", [], function() : void {
	$this->setFoo();
});
echo $this->getFoo();
```

> The result will be `bar` because the queries are ran in asynchronous, the code in the main thread will always <!-- Correct me if I'm wrong -->run faster than it.

Reviewer: ```suggestion
The result will be `bar` because the queries are run asynchronously. The code on the main thread will run faster than it.
```

Author: Huh "asynchronously" is a thing? Google translate shouted at me when I type it.

> To make the code gives out a correct result, you have to ensure `$this->setFoo()` runs before `echo $this->getFoo()`. The appropriate way is to put into the callback function, just like below:

Reviewer: ```suggestion
To make the code give a correct result, you have to ensure `$this->setFoo()` runs before `echo $this->getFoo()`. The appropriate way is to move `getFoo()` into the callback function, just like below:
```
